source/agora/consensus/validation/Block.d

```diff
@@ -112,6 +119,281 @@ public string isInvalidReason (const ref Block block, in ulong prev_height,
+
+    Params:
+        block = The genesis block to check
+
+    Returns:
+        `null` if the genesis block is valid, otherwise a string explaining
+        the reason it is invalid.
+
+*******************************************************************************/
+
+public string isGenesisBlockInvalidReason (const ref Block block) nothrow @safe
+{
+    if (block.header.height != 0)
+        return "GenesisBlock: The height of the block is not 0";
+
+    if (block.header.prev_block != Hash.init)
+        return "GenesisBlock: Header.prev_block is not empty";
+
+    if (block.txs.length == 0)
+        return "GenesisBlock: Transaction(s) are empty";
+
+    if (block.txs.length > Block.TxsInBlock)
+        return "GenesisBlock: The number of transactions is out of bounds";
+
+    if (!block.txs.isSorted())
+        return "GenesisBlock: Transactions are not sorted";
+
+    UTXOSetValue[Hash] utxo_set;
+    foreach (const ref tx; block.txs)
+    {
+        if (!(tx.type == TxType.Payment || tx.type == TxType.Freeze))
+            return "GenesisBlock: Transaction type allows only "
+                ~ "Payment and Freeze";
+
+        if (tx.inputs.length != 0)
+             return "GenesisBlock: Not empty input(s) in the transaction";
+
+        if (tx.outputs.length == 0)
+            return "GenesisBlock: No output(s) in the transaction";
+
+        Hash tx_hash = tx.hashFull();
+        foreach (idx, const ref output; tx.outputs)
+        {
+            // disallow negative amounts
+            if (!output.value.isValid())
+                return "GenesisBlock: Output(s) overflow or underflow"
+                    ~ "in the transaction";
+
+            // disallow 0 amount
+            if (output.value == Amount(0))
+                return "GenesisBlock: Value of output is 0"
+                    ~ "in the transaction";
+
+            const UTXOSetValue utxo_value = {
+                unlock_height: 0,
+                type: tx.type,
+                output: output
+            };
+            utxo_set[UTXOSetValue.getHash(tx_hash, idx)] = utxo_value;
+        }
+    }
+
+    Hash[] merkle_tree;
+    if (block.header.merkle_root !=
+        Block.buildMerkleTree(block.txs, merkle_tree))
+        return "GenesisBlock: Merkle root does not match header's";
+
+    // If there are no enrollments, return them here early.
+    if (block.header.enrollments.length == 0)
+        return null;
+
+    if (!isStrictlyMonotonic!"a.utxo_key < b.utxo_key"
+        (block.header.enrollments))
+        return "GenesisBlock: The enrollments should be arranged in "
+            ~ "ascending order by the utxo_key";
+
+    Set!Hash used_utxos;
+    bool findUTXO (Hash hash, size_t index, out UTXOSetValue value)
+        nothrow @safe
+    {
+        Hash utxo_hash;
+
+        if (index == size_t.max)
+            utxo_hash = hash;
+        else
+            utxo_hash = UTXOSetValue.getHash(hash, index);
```

Comment:
I think we only need the `size_t.max` case here, no ?

Reply:
That's right.